tools/parity_scale_decoder/src/main.rs

```diff
@@ -0,0 +1,279 @@
+//! Parity Scale decoder tool for Iroha data types. For usage run with `--help`
+
+#![allow(clippy::print_stdout, clippy::use_debug, clippy::unnecessary_wraps)]
+
+use std::{collections::BTreeMap, fmt::Debug, fs, io, io::Write, path::PathBuf};
+
+use clap::Parser;
+use eyre::{eyre, Result};
+use iroha_data_model::prelude::*;
+use parity_scale_codec::Decode;
+
+/// Parity Scale decoder tool for Iroha data types
+#[derive(Debug, Parser)]
+#[clap(version, about, author)]
+enum Args {
+    /// Show all available types
+    ListTypes,
+    /// Decode type from binary
+    Decode(DecodeArgs),
+}
+
+#[derive(Debug, clap::Args)]
+struct DecodeArgs {
+    /// Path to the binary with encoded Iroha structure
+    binary: PathBuf,
+    /// Type that is expected to be encoded in binary.
+    /// If not specified then a guess will be attempted
+    #[clap(short, long = "type")]
+    type_id: Option<String>,
+}
+
+/// Function pointer to [`DumpDecoded::dump_decoded()`]
+///
+/// Function pointer is used cause trait object can not be used
+/// due to [`Sized`] bound in [`Decode`] trait
+pub type DumpDecodedPtr = fn(&[u8], &mut dyn Write) -> Result<(), eyre::Error>;
```

Comment:
quite interesting approach. Reminded me a bit of functional interface in java

Reply:
though I think it is possible to generate an enum instead of a map in which case you wouldn't have to use function pointer. The enum would have quite a lot of variants but matching would probably be bit faster than finding a key in a map.

Reply:
AFAIK, in both cases the compiler is smart-enough to create a jump table.

Reply:
I'm not so sure, could be. That level of optimization hardly matters here, a map is fine. For future reference I just wanted to bring your attention to the fact that enum can be used here and I think it should be preferred

Reply:
I can't see how we can easily implement guessing with enum -- try each variant and print which is good

Reply:
It's not relevant here, it's good